subsys/net/lib/mqtt/mqtt.c

```diff
@@ -597,6 +597,24 @@ int mqtt_live(struct mqtt_client *client)
 	return 0;
 }
 
+u32_t mqtt_keepalive_time_left(const struct mqtt_client *client)
+{
+	u32_t elapsed_time = mqtt_elapsed_time_in_ms_get(
+					client->internal.last_activity);
+	u32_t keepalive_ms = 1000U * client->keepalive;
+
+	if (client->keepalive == 0) {
+		/* Keep alive not enabled. */
+		return UINT32_MAX;
+	}
+
+	if (keepalive_ms <= elapsed_time) {
```

Comment:
I think this will not work as expected if the clock has wrapped, you need to check the difference here instead.

Reply:
`elapsed_time` is already a difference (time since last MQTT activity, calculated in `mqtt_elapsed_time_in_ms_get()`), so we're not affected by system clock overflows here.

Reply:
Ok, I just checked last_activity variable and its documentation mentioned wall clock time.